File: src/TreeTraversal/lc0450_DeleteNodeinaBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from pkg.data_structure import TreeNode


# https://leetcode.cn/problems/delete-node-in-a-bst
class Solution:
    def deleteNode(self, root: TreeNode | None, key: int) -> TreeNode | None:

        def _pre_node(tree: TreeNode) -> TreeNode:  # 要求 tree 非空, 且 tree 一定存在 pre node
            node = tree.left
            while node.right is not None:
                node = node.right
            return node

        def _remove(tree: TreeNode) -> TreeNode | None:  # 要求 tree 非空
            if tree.left is None:
                return tree.right
            if tree.right is None:
                return tree.left
            pre = _pre_node(tree)
            pre.right = tree.right

            # no need to worry about the value of "tree", it will get collected when no reference points to it.
            # https://docs.python.org/zh-cn/3/library/sys.html#sys.getrefcount

            return tree.left

        def _dfs(root: TreeNode | None, key: int) -> TreeNode | None:
            if root is None:
                return None

            if key == root.val:
                return _remove(root)
            elif key < root.val:
                root.left = _dfs(root.left, key)
                return root
            else:  # key > root.val
                root.right = _dfs(root.right, key)
                return root

        return _dfs(root, key)

File: src/TreeTraversal/lc0450_DeleteNodeinaBST.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use crate::util::data_structure::TreeNode;
use std::cell::RefCell;
use std::rc::Rc;

pub struct Solution;

// https://leetcode.cn/problems/delete-node-in-a-bst
impl Solution {
    pub fn delete_node(
        root: Option<Rc<RefCell<TreeNode>>>,
        key: i32,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        fn _pre_node(tree: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
            // 要求 tree 非空, 且 tree 一定存在 pre node
            let mut node = tree.as_ref().unwrap().borrow().left.clone();
            while node.is_some() && node.as_ref().unwrap().borrow().right.is_some() {
                let temp = node.as_ref().unwrap().borrow().right.clone();
                node = temp
            }
            node
        }

        fn _remove(tree: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
            // 要求 tree 非空
            if tree.as_ref().unwrap().borrow().left.is_none() {
                return tree.as_ref().unwrap().borrow().right.clone();
            }
            if tree.as_ref().unwrap().borrow().right.is_none() {
                return tree.as_ref().unwrap().borrow().left.clone();
            }
            let pre = _pre_node(tree.clone());
            assert!(pre.is_some());
            pre.as_ref().unwrap().borrow_mut().right =
                tree.as_ref().unwrap().borrow().right.clone();
            tree.as_ref().unwrap().borrow().left.clone()
        }

        fn _dfs(root: Option<Rc<RefCell<TreeNode>>>, key: i32) -> Option<Rc<RefCell<TreeNode>>> {
            if root.is_none() {
                return None;
            }
            if key == root.as_ref().unwrap().borrow().val {
                _remove(root.clone())
            } else if key < root.as_ref().unwrap().borrow().val {
                let temp = _dfs(root.as_ref().unwrap().borrow().left.clone(), key);
                root.as_ref().unwrap().borrow_mut().left = temp;
                root
            } else {
                let temp = _dfs(root.as_ref().unwrap().borrow().right.clone(), key);
                root.as_ref().unwrap().borrow_mut().right = temp;
                root
            }
        }

        _dfs(root, key)
    }
}
